module/extension/Director/src/Director.cpp

```diff
@@ -0,0 +1,238 @@
+                auto& group   = groups[provider->type];
+
+                // Unbind any when state monitors
+                for (auto& when : provider->when) {
+                    when.handle.unbind();
+                }
+
+                // Erase the provider from this group
+                auto g_it = std::find(group.providers.begin(), group.providers.end(), provider);
+                if (g_it != group.providers.end()) {
+                    group.providers.erase(g_it);
+                }
+
+                // Now we need to deal with the cases where this provider was in use when it was unbound
+                if (provider->active) {
+                    if (group.providers.empty()) {
+                        // This is now an error, there are no providers to service the task
+                        log<NUClear::ERROR>(
+                            "The last provider for a type was removed while there were still tasks for it");
+                    }
+                    else {
+                        // TODO run the current task in the group to refresh which provider is chosen
+                        // run_task(group.active_task);
+                    }
+                }
+
+                // If there are no providers left in this group erase it too
+                if (group.providers.empty()) {
+                    groups.erase(provider->type);
+                }
+                // Erase from our list of providers
+                providers.erase(id);
+            }
+            else {
+                throw std::runtime_error("Attempted to remove a Provider that was not loaded");
+            }
+        }
+
+        Director::Director(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+
+            on<Configuration>("Director.yaml").then("Configure", [this](const Configuration& config) {
+                // clang-format off
+                std::string lvl = config["log_level"].as<std::string>();
+                if (lvl == "TRACE")      { this->log_level = NUClear::TRACE; }
+                else if (lvl == "DEBUG") { this->log_level = NUClear::DEBUG; }
+                else if (lvl == "INFO")  { this->log_level = NUClear::INFO;  }
+                else if (lvl == "WARN")  { this->log_level = NUClear::WARN;  }
+                else if (lvl == "ERROR") { this->log_level = NUClear::ERROR; }
+                else if (lvl == "FATAL") { this->log_level = NUClear::FATAL; }
+                // clang-format on
+            });
+
+            // Removes all the providers for a reaction when it is unbound
+            on<Trigger<Unbind>, Sync<Director>>().then("Remove Provider", [this](const Unbind& unbind) {  //
+                remove_provider(unbind.id);
+            });
+
+            // Add a provider provider
+            on<Trigger<ProvidesReaction>, Sync<Director>>().then("Add Provider", [this](const ProvidesReaction& r) {
+                add_provider(r.type, r.reaction, r.action);
```

Comment:
have you considered making `add_provider` take the `ProvidesReaction` object directly?

Reply:
I did, but I decided I didn't want to do that as that makes it a specific function. I'd rather it this way in case I ever want to add a provider separate to these reactions

Reply:
However now that I think about it I think I will change it to that